### ISSUE TITLE
Auth failing when using PBKDF2 SHA 256 hashing for password.

### ISSUE THREAD

**Issue by IcyBridge** [emqx/emqx-auth-redis#138]
Redis plugin setup with config

auth.redis.password_hash = pbkdf2,sha256,959,20

auth.redis.auth_cmd = HMGET mqtt_user:%u password salt

Somehow password hash with salt does not match those generated with C# program.

hash and salt is generated in c# following way:

using System.Security.Cryptography;

private static byte[] GetPbkdf2Bytes(string password, byte[] salt, int iterations, int outputBytes)
{
     var pbkdf2 = new Rfc2898DeriveBytes(password, 
     salt,iterations,HashAlgorithmName.SHA256);
     return pbkdf2.GetBytes(outputBytes);
}

after that hash to is converter to base64 datastring before saving to database and redis. Also salt is converted to base64 datastring.

Example hash generated from password "Example":

959:rzqDVn5QpFI9BPH1P/jWOXb49l4dCG/q:DPcqm/30krYVZmJi2lqyCA6Xbm8=

{iterations}:{salt}:{hash}

Iterations : 959
Salt: rzqDVn5QpFI9BPH1P/jWOXb49l4dCG/q
Hash: DPcqm/30krYVZmJi2lqyCA6Xbm8=

Hash request length is 20 bytes

2019-10-30 10:27:48.748 [error] jt2LU_SSHhbW2rB@10.211.113.4:64810 [Redis] Auth from redis failed: not_authorized

How should hash be generated to match with redis plugin ?

**Comment by tigercl** [comment]
Hi, @IcyBridge. You should save original data string rather than base64 data string to database.

**Comment by IcyBridge** [comment]
Thanks for tip! I tried many things, but still its not working. C# Pbkdf2 produces byte array.  Tried to convert byte array to following string types UTF8, ASCII, Hex("A45501FF..."), Unicode and byte array("\x09:y\xAB\xA7\xD..") without luck. What kind of data string is  plugin expecting for salt and hash ? Is it possible use some kind of binary string  

In the test file of the plugin, there was following line: {"mqtt_user:pbkdf2_password", ["password", "cdedb5281bb2f801565a1122b2563515", "salt", "ATHENA.MIT.EDUraeburn", "is_superuser", "0"]},

password hash look like hex string and salt looks like plain text. My case both hash and salt is originally bytes.  Is it possible to get basic example from datastring format for hash and salt? Thank you for your time!
